### PR TITLE
Explicity specify precision for all shader variables

### PR DESCRIPTION
**Pull request by nyquillerium**
According to [these](https://stackoverflow.com/questions/39707519/does-gles-mediump-float-have-enough-precision-for-mapping-display-coordinates) [answers](https://stackoverflow.com/questions/13780609/what-does-precision-mediump-float-mean/13780779#13780779), it is generally recommended to apply highp to operations that rely on position as they might not be sufficient for 10-bit float precision.

This is likely what is causing the issues in results, flashlight, and spinner glow, as this change fixes the display of those textures on both my device and a local emulator. However, this would likely need further testing to figure out whether or not this would be enough for all platforms.

### CODE REVIEW

File: osu.Framework/Resources/Shaders/sh_TextureRounded.fs [hunk 1]
```diff
@@ -1,43 +1,53 @@
 #ifdef GL_ES
 	precision mediump float;
+
+	// Not all GPUs support high precision, so we should fall back to medium precision.
+    #ifndef GL_FRAGMENT_PRECISION_HIGH
```
Comment (peppy) [inline osu.Framework/Resources/Shaders/sh_TextureRounded.fs]: .

**Comment by smoogipoo** [comment]
Please do the other shaders also. Consider doing something like what `sh_Backbuffer_Internal.sh` does, where it's automatically included into every vertex shader (except include it into _every_ shader instead).

I'm also slightly concerned about `sh_Backbuffer_Internal.sh` itself, which should be using as high precision for the attribute as possible.

**Comment by nyquillerium** [comment]
> I'm also slightly concerned about `sh_Backbuffer_Internal.sh` itself, which should be using as high precision for the attribute as possible.
In regards to `sh_BackBuffer_Internal`, [floats are already high precision for vertex shaders](https://stackoverflow.com/questions/5366416/in-opengl-es-2-0-glsl-where-do-you-need-precision-specifiers), but I guess we should define it anyways.

File: osu.Framework/Graphics/Shaders/ShaderPart.cs [hunk 2]
```diff
@@ -33,7 +33,7 @@ internal class ShaderPart
         private readonly List<string> shaderCodes = new List<string>();
 
         private readonly Regex includeRegex = new Regex("^\\s*#\\s*include\\s+[\"<](.*)[\">]");
-        private readonly Regex shaderInputRegex = new Regex("^\\s*(?>attribute|in)\\s+[^\\s]+\\s+([^;]+);");
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Shaders/ShaderPart.cs]: This is a very weird regex to me, the `.` is too greedy. Try something like:

`^\\s*(?>attribute|in)\\s+(?:(?:lowp|mediump|highp)\\s+)?([^\\s]+)\\s+([^\\s;]+)\\s*;`

Tests:
```
attribute lowp float name;
attribute mediump float name;
attribute highp float name;
attribute float name;
    attribute    float    name   ;
```

Comment (smoogipoo) [reply osu.Framework/Graphics/Shaders/ShaderPart.cs]: Update, a little more terse:

`^\\s*(?>attribute|in)\\s+(?:(?:lowp|mediump|highp)\\s+)?\\w+\\s+(\\w+)`

File: osu.Framework/Resources/Shaders/sh_Texture2D.vs [hunk 3]
```diff
@@ -1,24 +1,24 @@
...
 
-attribute vec2 m_Position;
-attribute vec4 m_Colour;
-attribute vec2 m_TexCoord;
-attribute vec4 m_TexRect;
-attribute vec2 m_BlendRange;
+attribute highp vec2 m_Position;
+attribute lowp vec4 m_Colour;
+attribute mediump vec2 m_TexCoord;
+attribute highp vec4 m_TexRect;
```
Comment (smoogipoo) [inline osu.Framework/Resources/Shaders/sh_Texture2D.vs]:
```suggestion
attribute mediump vec4 m_TexRect;
```

File: osu.Framework/Resources/Shaders/sh_Texture2D.vs [hunk 4]
```diff
@@ -1,24 +1,24 @@
...
+attribute highp vec2 m_BlendRange;
```
Comment (smoogipoo) [inline osu.Framework/Resources/Shaders/sh_Texture2D.vs]:
```suggestion
attribute mediump vec2 m_BlendRange;
```

File: osu.Framework/Resources/Shaders/sh_Texture2D.vs [hunk 5]
```diff
@@ -1,24 +1,24 @@
...
 
-varying vec2 v_MaskingPosition;
-varying vec4 v_Colour;
-varying vec2 v_TexCoord;
-varying vec4 v_TexRect;
-varying vec2 v_BlendRange;
+varying lowp vec2 v_MaskingPosition;
```
Comment (smoogipoo) [inline osu.Framework/Resources/Shaders/sh_Texture2D.vs]:
```suggestion
varying highp vec2 v_MaskingPosition;
```

File: osu.Framework/Resources/Shaders/sh_Texture2D.vs [hunk 6]
```diff
@@ -1,24 +1,24 @@
...
+varying mediump vec4 v_Colour;
```
Comment (smoogipoo) [inline osu.Framework/Resources/Shaders/sh_Texture2D.vs]:
```suggestion
varying lowp vec4 v_Colour;
```

File: osu.Framework/Resources/Shaders/sh_Texture2D.vs [hunk 7]
```diff
@@ -1,24 +1,24 @@
...
+varying mediump vec2 v_TexCoord;
+varying highp vec4 v_TexRect;
```
Comment (smoogipoo) [inline osu.Framework/Resources/Shaders/sh_Texture2D.vs]:
```suggestion
varying mediump vec4 v_TexRect;
```

File: osu.Framework/Resources/Shaders/sh_Texture2D.vs [hunk 8]
```diff
@@ -1,24 +1,24 @@
...
+varying highp vec2 v_BlendRange;
```
Comment (smoogipoo) [inline osu.Framework/Resources/Shaders/sh_Texture2D.vs]:
```suggestion
varying mediump vec2 v_BlendRange;
```

File: osu.Framework/Resources/Shaders/sh_Texture3D.vs [hunk 9]
```diff
@@ -1,14 +1,14 @@
...
 
-varying vec2 v_MaskingPosition;
-varying vec4 v_Colour;
-varying vec2 v_TexCoord;
-varying vec4 v_TexRect;
-varying vec2 v_BlendRange;
+varying highp vec2 v_MaskingPosition;
+varying lowp vec4 v_Colour;
+varying mediump vec2 v_TexCoord;
+varying highp vec4 v_TexRect;
```
Comment (smoogipoo) [inline osu.Framework/Resources/Shaders/sh_Texture3D.vs]:
```suggestion
varying mediump vec4 v_TexRect;
```

File: osu.Framework/Resources/Shaders/sh_Texture3D.vs [hunk 10]
```diff
@@ -1,14 +1,14 @@
...
+varying highp vec2 v_BlendRange;
```
Comment (smoogipoo) [inline osu.Framework/Resources/Shaders/sh_Texture3D.vs]:
```suggestion
varying mediump vec2 v_BlendRange;
```

File: osu.Framework/Resources/Shaders/sh_TextureRounded.fs [hunk 11]
```diff
@@ -1,43 +1,39 @@
...
 
-varying vec2 v_MaskingPosition;
-varying vec4 v_Colour;
-varying vec2 v_TexCoord;
-varying vec4 v_TexRect;
-varying vec2 v_BlendRange;
+varying highp vec2 v_MaskingPosition;
+varying lowp vec4 v_Colour;
+varying mediump vec2 v_TexCoord;
+varying highp vec4 v_TexRect;
```
Comment (smoogipoo) [inline osu.Framework/Resources/Shaders/sh_TextureRounded.fs]:
```suggestion
varying mediump vec4 v_TexRect;
```

File: osu.Framework/Resources/Shaders/sh_TextureRounded.fs [hunk 12]
```diff
@@ -1,43 +1,39 @@
...
+varying highp vec2 v_BlendRange;
```
Comment (smoogipoo) [inline osu.Framework/Resources/Shaders/sh_TextureRounded.fs]:
```suggestion
varying mediump vec2 v_BlendRange;
```

File: osu.Framework.Tests/IO/ShaderRegexTest.cs [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        public void TestComment()
+        {
+            const string test_string = "// The following implementation using mix and step may be faster, but stackoverflow indicates it is in fact a lot slower on some GPUs.";
+            performInvalidAttributeTest(test_string);
+        }
+
+        [Test]
+        public void TestNonAttribute()
+        {
+            const string test_string = "varying vec3 name";
```
Comment (smoogipoo) [inline osu.Framework.Tests/IO/ShaderRegexTest.cs]:
```suggestion
            const string test_string = "varying vec3 name;";
```

File: osu.Framework.Tests/IO/ShaderRegexTest.cs [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        public void TestSpacedAttribute()
+        {
+            const string test_string = "    attribute    float    name   ;";
+            performValidAttributeTest(test_string);
+        }
+
+        [Test]
+        public void TestNoPrecisionQualifier()
+        {
+            const string test_string = "attribute float name";
```
Comment (smoogipoo) [inline osu.Framework.Tests/IO/ShaderRegexTest.cs]:
```suggestion
            const string test_string = "attribute float name;";
```

File: osu.Framework.Tests/IO/ShaderRegexTest.cs [hunk 15]
```diff
@@ -0,0 +1,65 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Text.RegularExpressions;
+using NUnit.Framework;
+using osu.Framework.Graphics.Shaders;
+
+namespace osu.Framework.Tests.IO
```
Comment (smoogipoo) [inline osu.Framework.Tests/IO/ShaderRegexTest.cs]: Make a new `Graphics` namespace.

**Pull request merged**